### PR TITLE
Add positive tail length to tracks layer

### PR DESCRIPTION
**Pull request by JoOkuma**
# Description

This PR adds a new feature to visualize positive (forward in time) tracks.
This closes issue #3093 where it was further discussed.

positive_tail_track

I'm open to suggestions regarding the testing of this feature. I'm not very familiar with testing vispy shaders.

I called the forward tracks' length as `head_length` in the code.
Maybe I should revise that so it is more similar to the UI slider labels.

## Type of change
- [X] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

# References
Closes issue #3093 

# How has this been tested?
- [ ] all tests pass with my change

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by alisterburt** [comment]
awesome @JoOkuma !! 

Aware this is still a WIP, my first comment would be that the difference in sensitivity between the two sliders is quite jarring, are the lengths expressed as percentages of total possible length? If so, maybe describing these lengths in absolute units would be a bit more intuitive

Do you think having a double range slider with a label at the center and a handle at each end, one for moving the tail, one for moving the head, would be any good? Something like this
image

If so we could try to make a custom [QRangeSlider](https://github.com/napari/superqt/blob/22beed76086b95df2dbc9c7633dc1f57eba4b736/superqt/sliders/_sliders.py#L34) to add a label in the middle and limit the handle positions, so they can't move past the midpoint, could be fun!

**Comment by JoOkuma** [comment]
I kept the previous behavior of the tail length, where you can increase it up to constant (300), but its effect on the tracks' colors is much more noticeable when changing the head length.

Your suggestion is great!
Do you think it would be better for the double range slider to range between - the number of time points to + the number of time points and we clip the values under the hood according to the current time point, or it could change its values dynamically, as the user moves forward in time it would have more tail length and less head length.

I think the second one is more elegant but it could be annoying to have the range scale changing all the time.

**Comment by alisterburt** [comment]
> Do you think it would be better for the double range slider to range between - the number of time points to + the number of time points and we clip the values under the hood according to the current time point

This feels natural to me, would be curious if anyone else prefers the dynamic suggestion though!

**Comment by tlambert03** [comment]
I don't know, I quite liked the idea of a range slider with a middle limit!  And +/- max/min would just be a reasonable tail length max in either direction?

**Comment by sofroniewn** [comment]
All sounds reasonable to me, but I havn't tried it out. I havn't used tracks so much myself. Maybe @kevinyamauchi or @jni can give this a review too and once it's got some approvals I'll give it a merge. Thanks!!

**Review by sofroniewn — APPROVED** [review]
LGTM! I'm fine with this addition, very clean. If one more person who uses tracks more then i do was able to give this a look over too that would be great, otherwise i'll give this a merge by end of day tomorrow. thanks for the contribution @JoOkuma and sorry it took a while to review!!

**Comment by kevinyamauchi** [comment]
I can have a look today. Sorry for the delay, @JoOkuma !

File: napari/_qt/layer_controls/qt_tracks_controls.py
Comment (kevinyamauchi) [inline napari/_qt/layer_controls/qt_tracks_controls.py]: It feels a bit odd to me that the clipping is done in the layer controls. This means that the controls and layer model can become out of sync (e.g., set the `head_length` value to greater than `MAX_TAIL_LENGTH` on the layer.

File: napari/_qt/layer_controls/qt_tracks_controls.py
Comment (kevinyamauchi) [inline napari/_qt/layer_controls/qt_tracks_controls.py]: I think the label should be the same as the property name on the layer.

File: napari/_vispy/_vispy_tracks_shader.py
Comment (kevinyamauchi) [inline napari/_vispy/_vispy_tracks_shader.py]: Is this space correct?

File: napari/_qt/layer_controls/qt_tracks_controls.py
Comment (kevinyamauchi) [reply napari/_qt/layer_controls/qt_tracks_controls.py]: I now see that this is the case for the `tail_length` already. I think the clipping should be done on the layer so that the layer and controls stay in sync. Thoughts? cc @sofroniewn

**Review by kevinyamauchi — COMMENTED** [review]
Thanks, @JoOkuma ! This looks great. There are just two minor things I think should be addressed before we can merge:

- we should call the positive tail the same thing in the layer control labels and on the layer model. I don't have a strong preference between "positive_tail_length" and "head_length", but I think it should be uniform. Any preference?
- I think we should consider clipping the head/tail length on the model rather than in the qt layer controls. I am concerned that clipping just in the layer controls allows the layer controls and model to get out of sync (see my comment below).

I think we are good to go once we decide on those two things! Thanks again!

File: napari/_qt/layer_controls/qt_tracks_controls.py
Comment (sofroniewn) [reply napari/_qt/layer_controls/qt_tracks_controls.py]: yeah - we've run into problems with this in other parts of the codebase with sliders and their max length - i think the paintbrush size slider was the last place. I can't remember right now what we did there though, does @ppwadhwa @DragaDoncila or @tlambert03 recall - i think some of you were involved in that work

Comment (tlambert03) [reply napari/_qt/layer_controls/qt_tracks_controls.py]: yeah, it's a bit of a bummer that our layers aren't pydantic models, ideally the declaration of `head_length` on the model would be a `conint` that contains this information, and widgets could then easily represent it (with a single source of truth).  Until then, I don't have an elegant solution... in other places, we've added annoying logic to allow the widget to "grow" it's range to include whatever value has been provided programatically.

In any case, I agree with @kevinyamauchi that keeping the model and the view in sync is the main priority.  If you clip it on the widget, you have to then clip it on the model too

**Comment by JoOkuma** [comment]
Thanks for the comments @kevinyamauchi.
I need to finish a few things at the Biohub, but this should be done by the end of the week.

File: napari/_qt/layer_controls/qt_tracks_controls.py
Comment (sofroniewn) [reply napari/_qt/layer_controls/qt_tracks_controls.py]: yeah, we can do a clip on model, though we might want to emit a warning

**Comment by JoOkuma** [comment]
I moved the clipping to the Tracks layer. The maximum tail width constant was changed to keep the previous behavior of multiplying by 2.0 (I don't why we need this though).

I decided to rename the '+ length' to 'head length' for consistency, but maybe 'neck length' would be a more appropriate choice 😂.

Regarding the sensitivity of the tracks fade to the tracks' length. I think it could be fixed in a later PR with the dynamic double ranged slider.

**Review by sofroniewn — APPROVED** [review]
Ok I will approve this now and give it a merge so it is in the next RC - i think longterm though we're going to want to avoid this sort of clipping in our models so that are sliders are ok - it sort of feels like putting the cart before the horse, or whatever the right expression is :-)

**Pull request merged**
